Migration of page Shop Parameters > Product Settings

sarjon opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrate Product Settings page to symfony
| Type?         | new feature
| Category?     | BO
| BC breaks?    | yes (overrides)
| Deprecations? | no?
| Fixed ticket? | n/a
| How to test?  | Go to Configure > Shop Parameters > Product Settings. Some form fields are multilingual, make sure to have at least 2 active languages to test all functionality.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8680)
<!-- Reviewable:end -->

Commit:
add allowed types for translate text type options

Commit:
move form submit buttons to the right side

Commit:
rearrange form blocks display

Commit:
updated locale select styling

Commit:
add missing product stock settings fields

Commit:
retrieve product stock settings from legacy

Commit:
make translate text inputs not required by default

Commit:
fix stock configuration for product

Commit:
add all translatable inputs toggle in form

Commit:
update configuration classes to save all form values

Commit:
move js translatable input component to class

Commit:
added js stock management option handler

Commit:
add display quantities handling in js stock management handler

Commit:
remove AdminPPreferencesController from tests

Commit:
delete AdminPPreferencesController

Commit:
add missing class docs for form types

Commit:
add helper method for employee redirection to default page

Commit:
update form icons

Commit:
fix StockConfiguration to return array when default values are not available

Commit:
remove unnecessary variable definition

Commit:
fix spacing in PageType

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/PaginationType.php:
```diff
@@ -0,0 +1,55 @@
+            ->add('products_per_page', IntegerType::class)
+            ->add('default_order_by', ChoiceType::class, [
+                'choices' => [
+                    'Product name' => 0,
```
be careful, this need to be translated in every language, you probably need to use the translator here and rely on TranslatorAwareType form instead of AbstractType

sarjon replied:
Actually i tried switching to another language and translations worked fine for these strings. I also checked other forms that a migrated and they implement choices the same way. Im wondering if the changes are necessary here?

eternoendless replied:
It's working for me too

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -0,0 +1,70 @@
+<?php
+
```
License is missing here

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php:
```diff
@@ -0,0 +1,53 @@
+<?php
+
```
License is missing here

mickaelandrieu commented on src/PrestaShopBundle/Resources/config/adapter/services.yml:
```diff
@@ -132,3 +132,23 @@ services:
         class: 'PrestaShop\PrestaShop\Adapter\Configuration\PhpParameters'
         arguments: ['%kernel.root_dir%/config/parameters.php']
+
+    prestashop.adapter.product_general.configruation:
```
```diff
- configruation
+ configuration
````

mickaelandrieu commented on src/PrestaShopBundle/Resources/config/services.yml:
```diff
@@ -515,3 +515,18 @@ services:
+    prestashop.admin.product_preferences.data_provider:
+        class: 'PrestaShopBundle\Form\Admin\ShopParameters\ProductPreferences\ProductPreferencesFormDataProvider'
+        arguments:
+            - '@prestashop.adapter.product_general.configruation'
```
needs also to be updated :)

mickaelandrieu commented on src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/product_preferences.html.twig:
```diff
@@ -0,0 +1,248 @@
+    <div class="container">
+        {{ form_start(form, {'attr': {'class': 'form'}}) }}
+        <div class="row">
+            {% block product_preferences_general %}
```
maybe can you create some template into `views/Admin/Configure/ShopParameters/Blocks` and use include? This make templates easier to be overridden and customized using bundles/modules.

It's not a strict rule, but since we've refactored Product Pages templates we try to keep template file under 100 lines :)

sarjon replied:
Good idea. :) Updated template.

mickaelandrieu commented on admin-dev/themes/new-theme/js/product-preferences-page/index.js:
```diff
@@ -0,0 +1,39 @@
+
+const $ = window.$;
+
+$(() => {
```
Sounds like this is specific to _product_preferences_ page, why not create a new Webpack entry named product_preference_page and include it into this page instead?

sarjon replied:
It sounds like a good idea. I have added new entry in webpack, but still im not sure if it is the best option. if we add new entry for every page, webpack config may get really big, but maybe it's not a problem. :)

sarjon replied:
I also wonder if we should create separate directory like 'pages' instead of suffixing each directory with '-page'. What do you think?

mickaelandrieu commented on src/Adapter/Product/StockConfiguration.php:
```diff
@@ -0,0 +1,81 @@
+<?php
+
```
License missing

sarjon replied:
Fixed.

mickaelandrieu commented on src/Adapter/Product/StockConfiguration.php:
```diff
@@ -0,0 +1,81 @@
+     */
+    private $configuration;
+
+    public function __construct(Configuration $configuration)
```
would you mind to rely on ConfigurationInterface instead? This one =>https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/ConfigurationInterface.php

Sorry if it's done with Configuration classe everywhere, this needs to be updated/improved (my bad).

sarjon replied:
Yes! :+1:  Fixed.

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/ProductPreferencesController.php:
```diff
@@ -0,0 +1,123 @@
+     */
+    public function indexAction(Request $request)
+    {
+        $legacyController = $request->attributes->get('_legacy_controller');
```
I like this, I can't remember why we need the constant ... I need to check we never rely on them, and if we don't it's safe to use your way to do... this make me think about using a controller_argument resolver to redirect to default page, what do you think?

sarjon replied:
I dont know if i understand your idea correctly, maybe you could give an example?

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -0,0 +1,94 @@
+        ;
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -0,0 +1,94 @@
+        ]);
+    }
+
+    public function getBlockPrefix()
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -0,0 +1,70 @@
+        ;
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -0,0 +1,70 @@
+        ]);
+    }
+
+    public function getBlockPrefix()
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/PaginationType.php:
```diff
@@ -0,0 +1,55 @@
+        ;
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/PaginationType.php:
```diff
@@ -0,0 +1,55 @@
+        ]);
+    }
+
+    public function getBlockPrefix()
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -0,0 +1,70 @@
+ */
+class StockType extends TranslatorAwareType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -0,0 +1,70 @@
+        ;
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -0,0 +1,70 @@
+        ]);
+    }
+
+    public function getBlockPrefix()
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php:
```diff
@@ -0,0 +1,53 @@
+ */
+class TranslateTextType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    public function buildView(FormView $view, FormInterface $form, array $options)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php:
```diff
@@ -0,0 +1,53 @@
+        $view->vars['hide_locales'] = 1 >= count($options['locales']);
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
```
{@inheritdoc}

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php:
```diff
@@ -0,0 +1,53 @@
+        $resolver->setAllowedTypes('options', 'array');
+    }
+
+    public function getBlockPrefix()
```
{@inheritdoc}

mickaelandrieu reviewed (dismissed):
Good job! Some minor comments 👍

sarjon commented:
thanks for review @mickaelandrieu! ill take a look at your comments tomorrow. :)

Commit:
fix missing licenses & docblocks

Commit:
split product preferences template into separate templates

Commit:
fix spelling in service naming

Commit:
update translation domains

Commit:
rely on interface in configuration classes

Commit:
add new webpack entry for product preferences page js

marionf commented:
Hello @sarjon 

When I disable stock management, I have the following issue when I try to add or edit a product: 

![capture du 2018-01-18 11-17-39](https://user-images.githubusercontent.com/13449658/35101261-2dd64788-fc5f-11e7-9904-dff6321fcd2b.png)


My back-office language is French and I have labels in English:

![capture du 2018-01-18 14-33-35](https://user-images.githubusercontent.com/13449658/35101309-6353617a-fc5f-11e7-9ef2-e7294479195d.png)

![capture du 2018-01-18 14-33-46](https://user-images.githubusercontent.com/13449658/35101315-684e05ae-fc5f-11e7-8c52-457ea3de5ef5.png)


All the tooltips are also in English:

![capture du 2018-01-18 14-35-21](https://user-images.githubusercontent.com/13449658/35101332-79917800-fc5f-11e7-9c00-829190562682.png)

The "Delivery time of in-stock products" and "Delivery time of out-of-stock products with allowed backorders" are not displayed on the product page (Front-office)

![capture du 2018-01-18 11-32-18](https://user-images.githubusercontent.com/13449658/35101438-d7f00d62-fc5f-11e7-820d-d21be299d7b7.png)

![capture du 2018-01-18 11-33-24](https://user-images.githubusercontent.com/13449658/35101449-df5ecc78-fc5f-11e7-97d9-3b482642ac50.png)

![capture du 2018-01-18 11-45-19](https://user-images.githubusercontent.com/13449658/35101458-e79d2006-fc5f-11e7-95da-07234c246a68.png)

![capture du 2018-01-18 11-45-48](https://user-images.githubusercontent.com/13449658/35101463-ebbc5b66-fc5f-11e7-8474-245702e0cad2.png)


The delivery time are not displayed on the product page (Back-office), instead I have this:

![capture du 2018-01-18 12-12-07](https://user-images.githubusercontent.com/13449658/35101507-1d53ad14-fc60-11e7-9553-2b7cfc0606df.png)

sarjon commented:
hey @marionf thanks for your review! However, I think that only translations issue are related to this PR. Here's my thoughts on this:

1. I tried to disable Stock management on develop branch (without my changes) and I get the same error. I think it's not related to this PR, can you confirm?

2. 'Delivery time of in-stock products' and 'Delivery time of out-of-stock products with allowed backorders' are actually displayed in Front Office product page. You have to select option 0, 1 or 2 under BO Product shipping tab 'Delivery Time' to be able to see it. You can reproduce it on develop branch as well. I think it's not related to this PR, can you confirm?

marionf commented:
Sorry my develop branch wasn't up to date, you're right I have them also on develop.
So, we have just the language issue :)

Commit:
update translation text & domains

sarjon commented:
thanks for your quick reply @marionf. I changed some text in migrated controller but now I have updated everything with original text & domains from legacy and it seems to be working now. :)

marionf commented:
It looks good to me, thank you @sarjon :)

Quetzacoalt91 commented on src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Class is responsible for creating translatable text inputs
+ */
+class TranslateTextType extends AbstractType
```
Why did you make a new FormType independent from the existing `PrestaShopBundle\Form\Admin\Type\TranslateType`?
They seem to be quite related, but they do not have code in common.

sarjon replied:
Hmmm, thats a good point @Quetzacoalt91. Is it used to render textareas? 
Maybe it to would be the best to have TranslateTextType & TranslateTextareaType which shares common code?

Quetzacoalt91 replied:
Not exactly. It seems to display the language selector. It must be combined with the Type you want to display:
### Text:
https://github.com/PrestaShop/PrestaShop/blob/3f9808dd5b53c75529ec9243f00bdba9857cc594/src/PrestaShopBundle/Form/Admin/Product/ProductInformation.php#L143-L158

### TextArea with TinyMce (prior #8604 ):
https://github.com/PrestaShop/PrestaShop/blob/3f9808dd5b53c75529ec9243f00bdba9857cc594/src/PrestaShopBundle/Form/Admin/Product/ProductInformation.php#L159-L177

### TextArea with TinyMce (from #8604):
https://github.com/Quetzacoalt91/PrestaShop/blob/8a5f67d5e2651de8a15e3a875fc51707cd0496cd/src/PrestaShopBundle/Form/Admin/ShopParameters/General/MaintenanceType.php#L58-L66

It seems you just had to adapt the existing template in order to display the selector properly.
However, this requires some check on the Product page, where these inputs already exist.

sarjon replied:
ok, thanks for clarification. :) However, I'm no sure if it's a good idea to have 'all in one' type. Problem I see is that you have to specify which type you want to combine, does this mean you can combine ChoiceType, DateTimeType or any other type available in Symfony? It's confusing. I would like to see different types for different inputs, like Symfony does it with TextType, TextareaType and others so you can add it to your form without any addtional options. For us it means we dont have to adapt TranslateType template for many different cases. It's just my thoughts. What do you think @Quetzacoalt91?

Quetzacoalt91 replied:
I haven't tested, but I do not see why we could not use the other Symfony types.
As you can see on the 2 first lines, we always have the choice to use the input type we want.

... or I misunderstood what you meant?

sarjon replied:
yeah, my idea was a bit different.

do you think i should adapt TranslateType and use it in this page?

Quetzacoalt91 replied:
Basically we have a duplication of code here between `PrestaShopBundle\Form\Admin\Type\TranslateType` && `PrestaShopBundle\Form\Admin\Type\TranslateTextType`.

Let's keep your solution as well as you made a specific integration for this page. But we **will have to** merge these classes prior the 1.7.4.0 release. The objective is to have the same behavior everywhere. If we need to adapt the existing pages, we will.

Quetzacoalt91 replied:
Can be done on another PR. The task has been created: http://forge.prestashop.com/browse/BOGOSS-87

sarjon replied:
thanks @Quetzacoalt91. Unfortunately I cannot see the task, I get 404 even though i'm logged in.

Commit:
fix spacing in PageType

eternoendless commented:
Thanks for this contribution @sarjon !

Please keep in mind that PrestaShop follows [Airbnb's javascript guidelines](https://github.com/airbnb/javascript), which requires [using 2 spaces for indentation instead of 4](https://github.com/airbnb/javascript#whitespace--spaces) in javascript files.

---

Reviewed 22 of 38 files at r1, 16 of 23 files at r2, 6 of 6 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[admin-dev/themes/new-theme/js/product-preferences-page/index.js, line 31 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L33j1dK4G6n6WwzzAhJ-r1-31:-L40Z8ZaQi0VJVMd9ulC:b-6lb4au) ([raw file](https://github.com/prestashop/prestashop/blob/72a4f5bcdf1758686a5d80126979dcbd6f67b6d2/admin-dev/themes/new-theme/js/product-preferences-page/index.js#L31)):*
> if we add new entry for every page, webpack config may get really big

I don't think that will be a problem

> I also wonder if we should create separate directory like 'pages' instead of suffixing each directory with '-page'. What do you think?

It's a good idea, we already have that for Vue pages (new-theme/js/app/pages), but since we're putting them in hold for a while, it would be nice to follow the same structure. Go ahead!

---

*[admin-dev/themes/new-theme/scss/components/_translatable-input.scss, line 3 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L40aOM-UNIAdPp93qrC:-L40aOM-UNIAdPp93qrD:b-sdap0e) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/admin-dev/themes/new-theme/scss/components/_translatable-input.scss#L3)):*
> ```SCSS
> .locale-input-group {
>   .dropdown {
>     margin-left: 10px;
> ```

Please use REMs instead of pixels. Here's a handy calculator which converts to EMs but follows the same principle: http://pxtoem.com/

---

*[admin-dev/themes/new-theme/scss/components/_translatable-input.scss, line 6 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L40aXZZD1KH56UOmARD:-L40aXZZD1KH56UOmARE:b7jn4uf) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/admin-dev/themes/new-theme/scss/components/_translatable-input.scss#L6)):*
> ```SCSS
> 
>     button.dropdown-toggle {
>       border-color: #bbcdd2;
> ```

Please use variables: `$gray-light` (you can find them in admin-dev/themes/new-theme/scss/config/_settings.scss)

---

*[src/Adapter/Product/GeneralConfiguration.php, line 76 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L40yovu_0rLEh9KXCQq:-L40yovu_0rLEh9KXCQr:bidwvc2) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/Adapter/Product/GeneralConfiguration.php#L76)):*
> ```PHP
>             $this->configuration->set('PS_QTY_DISCOUNT_ON_COMBINATION', (int) $config['quantity_discount']);
>             $this->configuration->set('PS_FORCE_FRIENDLY_PRODUCT', (int) $config['force_friendly_url']);
>             $this->configuration->set('PS_PRODUCT_ACTIVATION_DEFAULT', (int) $config['default_status']);
> ```

You can also chain `set` calls:

```php
$this->configuration
    ->set(...)
    ->set(...)
```

---

*[src/Adapter/Product/PageConfiguration.php, line 33 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L41-690MxaZ-2dMPlW7:-L41-690MxaZ-2dMPlW8:b-silxkw) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/Adapter/Product/PageConfiguration.php#L33)):*
> ```PHP
> use Symfony\Component\OptionsResolver\OptionsResolver;
> 
> class PageConfiguration implements DataConfigurationInterface
> ```

Please add a description for this class

---

*[src/Adapter/Product/PaginationConfiguration.php, line 33 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L41-DiteC1x_NHY_MNT:-L41-DiteC1x_NHY_MNU:b-silxkw) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/Adapter/Product/PaginationConfiguration.php#L33)):*
> ```PHP
> use Symfony\Component\OptionsResolver\OptionsResolver;
> 
> class PaginationConfiguration implements DataConfigurationInterface
> ```

Please add a description for this class

---

*[src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php, line 92 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L414gHjduzX9ZRVphHh:-L414gHjduzX9ZRVphHi:b-x4jint) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php#L92)):*
> ```PHP
>     {
>         $this->cacheClearer->clearSmartyCache();
>         $this->cacheClearer->clearMediaCache();
> ```

I'm under the impression that this wasn't always done before

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-:-L40_PmBi8m-TIAq2l0z:bkp9bmy)*
<!-- Sent from Reviewable.io -->

Commit:
move controller scripts to pages directory

Commit:
use 2 spaces indentation instead of 4 in javascript files

Commit:
use em instead of px in translatable input scss component

Commit:
use scss variable instead of hardcoded value

Commit:
add missing doc blocks

sarjon commented:
*[src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php, line 92 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L414gHjduzX9ZRVphHh:-L46GOuKX2bfO7Iy0jrj:bpkwb0v) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php#L92)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

I'm under the impression that this wasn't always done before
</blockquote></details>

Actually it was always executed before options were saved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8680)*
<!-- Sent from Reviewable.io -->

eternoendless commented:
Reviewed 16 of 16 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php, line 92 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L414gHjduzX9ZRVphHh:-L46v-BU-BLjbMT60tnU:bxi5fdb) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php#L92)):*
<details><summary><i>Previously, sarjon (Šarūnas Jonušas) wrote…</i></summary><blockquote>

Actually it was always executed before options were saved.
</blockquote></details>

To me it looks like it only did that when disabling stock management, only if it was active before:

https://github.com/PrestaShop/PrestaShop/blob/1a897064c6207e8f66a510f428a77f5052a80c38/controllers/admin/AdminPPreferencesController.php#L285

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8680)*
<!-- Sent from Reviewable.io -->

sarjon commented:
*[src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php, line 92 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L414gHjduzX9ZRVphHh:-L477hRMfl9jdKAyqDcB:b-yom57q) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php#L92)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

To me it looks like it only did that when disabling stock management, only if it was active before:

https://github.com/PrestaShop/PrestaShop/blob/1a897064c6207e8f66a510f428a77f5052a80c38/controllers/admin/AdminPPreferencesController.php#L285

</blockquote></details>

It is a bit difficult to see, but this if statement ends at line 299.

https://github.com/PrestaShop/PrestaShop/blob/1a897064c6207e8f66a510f428a77f5052a80c38/controllers/admin/AdminPPreferencesController.php#L299

Another if statement starts here, however it always evalues to true.

https://github.com/PrestaShop/PrestaShop/blob/1a897064c6207e8f66a510f428a77f5052a80c38/controllers/admin/AdminPPreferencesController.php#L301

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8680)*
<!-- Sent from Reviewable.io -->

eternoendless commented:
Looks good to me! It just needs to be rebased in order to fix the merge conflicts

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php, line 92 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L414gHjduzX9ZRVphHh:-L4kHYZFFekqhzJVpEMT:b-qce5g7) ([raw file](https://github.com/prestashop/prestashop/blob/e1195db13062fdc5974966890e284dc800aaed23/src/PrestaShopBundle/Form/Admin/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php#L92)):*
<details><summary><i>Previously, sarjon (Šarūnas Jonušas) wrote…</i></summary><blockquote>

It is a bit difficult to see, but this if statement ends at line 299.

https://github.com/PrestaShop/PrestaShop/blob/1a897064c6207e8f66a510f428a77f5052a80c38/controllers/admin/AdminPPreferencesController.php#L299

Another if statement starts here, however it always evalues to true.

https://github.com/PrestaShop/PrestaShop/blob/1a897064c6207e8f66a510f428a77f5052a80c38/controllers/admin/AdminPPreferencesController.php#L301
</blockquote></details>

You're right! My mistake.

---

*[src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php, line 39 at r3](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-L33j1dK4G6n6WwzzAhi-r3-39:-L4kJtHaXwedC_J2ZmH5:b-s9ox7a) ([raw file](https://github.com/prestashop/prestashop/blob/5cb9024255fc1367df98b9de02ab1b93a2e96e67/src/PrestaShopBundle/Form/Admin/Type/TranslateTextType.php#L39)):*
<details><summary><i>Previously, sarjon (Šarūnas Jonušas) wrote…</i></summary><blockquote>

thanks @Quetzacoalt91. Unfortunately I cannot see the task, I get 404 even though i'm logged in.
</blockquote></details>

Don't worry @sarjon, we'll take care of it :thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-:-L4kKNH2mzDMA4I6PGfs:b-orp64d)*
<!-- Sent from Reviewable.io -->

Commit:
resolve conflict

Commit:
resolve conflict

sarjon commented:
@eternoendless rebased.

eternoendless commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 11 of 12 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-:-L4pvgBGF5elc_2md1Mt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kpodemski commented:
Guys, the main issue i have is that there's literally zero option for customization of the core Configuration pages right now, because of migration to Symfony instead of doing further steps to provide more flexibility, we have more hardcoded things in code

Example? In bigger projects sometimes we need to rework some configuration pages, we can't just hide the tab and create everything from scratch, also, this was never a good idea, the thing is that in 1.6 we had a possibility to override files, right now we don't have it.

Another example? Settigns which depends on each other, in a solutions like ReduxFramework for WordPress or in many PrestaShop themes there are features to create a field which depends on other field value, you can toggle visiblity on value etc. and this all in a array like this:

```
[
'name' => 'something',
'depends_on' => 'other_field:valueofthisfield',
]
```

or

```
[
'name' => 'something',
'js_callback' => 'someFunction', // this will run a js function with a value as a param, i pretty sure this was already implemented in 1.6
]
```

Don't get me wrong, i love the idea of using Symfony, but lets do it right, lets make this software as flexible as it should be and not wait till 1.8 :/

Please make sure to treat all array of the fields as dynamic as possible, we can have a hook here and there, or some other way to work with forms in Configuration pages, @mickaelandrieu any thoughts?

eternoendless commented:
Hi @kpodemski 

I hear your concern about customization. We are in the middle of the transition to Symfony and I agree that it's a bit... messy, for lack of a better word.

Extensibility is a big subject and we have a general idea of how we want to do it in the long term (I'm talking probably 1 or 2 major versions from now): the BO should be all-Vue on the front-end, extensible through javascript APIs, while the back-end should stop building forms in html.

_How_ we'll get there is unclear right now, but we are working on it.

In the meantime, this PR should help in adding fields to forms starting on 1.7.4: https://github.com/PrestaShop/PrestaShop/pull/8368

kpodemski commented:
Hi @eternoendless 

i'm glad that we share same concerns about current situation

btw. all-Vue BO... i'm in love already =)

overall, keep good work guys, i belive that the last time this repository had so much good stuff in Pull Requests was few years ago :D

mickaelandrieu commented:
Hi @kpodemski,

you can override almost everything like in 1.6... even better than in 1.6 because you don't have to override the controller:

* vues are overridable in modules mostly the same way than 1.6 but this time you can override only 1 block (this is why you have so much named blocks in the new templates and why I've refactored the Product Pages templates). Data providers are out of controllers and overridable in modules, allowing you to pass the data you need when you need it without alter PrestaShop code so the upgrade are easier

* forms: right now you can create your owns and alter the provided ones, by declaring them in modules and/or using evdnt system of Symfony, as you know I'm already working on a new DX improvements on this specific forms customisations but I want to take my time and play a little bit with it first (I guess you have reviewed my PR about it, I haven't give up, I'm in hollidays)

So, please don't worry! What is missing here is not the features but the docs and we are working on it.

Thanks for your support, this is for people like you we want to make our best!

Cya!

eternoendless commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8680#-:-L5rpKLP_pXXnmDIXPLL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

marionf commented:
Hello @sarjon 

Before I didn't have the opportunity to enter negative numbers for the following fields:
- Products per page
- Number of days for which the product is considered 'new' 
- Display remaining quantities when the quantity is lower than

![capture du 2018-03-14 16-41-32](https://user-images.githubusercontent.com/13449658/37417341-4cc5dd6c-27b0-11e8-85a1-bd7aecd01ca7.png)

Now I can:

![capture du 2018-03-14 16-40-12](https://user-images.githubusercontent.com/13449658/37417362-5d992e78-27b0-11e8-9dc2-3b4ab216e6d9.png)

It causes an issue in FO, especially for products per page:

![capture du 2018-03-14 17-34-27](https://user-images.githubusercontent.com/13449658/37417532-d5c64b42-27b0-11e8-9493-8633dcb24bd3.png)

Can you add an error message like before with the name of the field which is wrong ?

And according to our UI Kit (https://cdn.rawgit.com/PrestaShop/prestashop-ui-kit/master/index.html#forms), can you use the custom-select class ?

![capture du 2018-03-14 17-10-29](https://user-images.githubusercontent.com/13449658/37417719-3c4a9ec2-27b1-11e8-9d35-48e8643dfd90.png)

Thanks :)

Commit:
update choice_widget_collapsed to use custom-select class instead of form-control

sarjon commented:
hey @eternoendless can you check https://github.com/PrestaShop/PrestaShop/pull/8680/commits/aa347c6e794ec24d82b19c5eb6db6ce45f4f8adf commit as it affects all forms. It uses custom-select instead of form-control (according to UI kit) and fixes issue of letters being cut down.
Before:
![before-custom-select](https://user-images.githubusercontent.com/15143151/37509847-bb73414e-2901-11e8-93ff-8efafcb57c6a.png)

After:
![after-custom-select](https://user-images.githubusercontent.com/15143151/37509854-c1536260-2901-11e8-8fe6-43a9eb5cbf49.png)

mickaelandrieu commented:
> hey @eternoendless can you check aa347c6 commit as it affects all forms. It uses custom-select instead of form-control (according to UI kit) and fixes issue of letters being cut down.

Hi @sarjon,

It's not "instead", it's in addition to "form-control". Can you update your commit accordingly?

Thank you

sarjon commented:
thanks for reply @mickaelandrieu, you mean they both (custom-select & form-control) should be present on select tag?

mickaelandrieu commented:
@sarjon yup!

![yep](https://user-images.githubusercontent.com/1247388/37527073-64b5daf8-2931-11e8-93a2-c368e1fcff40.png)

sarjon commented:
okay @mickaelandrieu , but using both custom-select and form-control cuts down letters, should I not worry about it?

mickaelandrieu commented:
@sarjon I'm worried about how the forms will react using mobile or tablet display without the classes that come from bootstrap 4, wdyt @eternoendless?

eternoendless commented:
`custom-select` is actually from bootstrap, so you shouldn't worry about responsiveness: https://getbootstrap.com/docs/4.0/components/input-group/#custom-select

We should use `custom-select` only to ensure cross-browser consistency.

mickaelandrieu commented:
@sarjon ok forget about my comment, you were right :innocent: 

Need a rebase then LGTM?

Commit:
add validation & validation messages for some fields

sarjon commented:
@mickaelandrieu @eternoendless can you review https://github.com/PrestaShop/PrestaShop/pull/8680/commits/c695b352574d217cbc248f3c14be9e8b92023aee commit? I'm not really happy with current validation implementation, but maybe you have better ideas?

Just an idea, shouldn't form handler be aware of Request and Form instance? So we could move this boilerplate code from controller to form handler?

```php
        $formHandler = $this->get('prestashop.admin.product_preferences.form_handler');

        $form = $formHandler->getForm();
        $form->handleRequest($request);

        if ($form->isSubmitted()) {
            $data = $form->getData();

            $errors = $formHandler->save($data);

            if (0 === count($errors)) {
                $this->addFlash('success', $this->trans('Update successful', 'Admin.Notifications.Success'));
            } else {
                $this->flashErrors($errors);
            }
        }
```
So it could become something like this in controller:
```php
$formHandler = $this->get('prestashop.admin.product_preferences.form_handler');
$someResult = $formHandler->handle($request);

if ($errors = $someResult->getErrors()) {
     $this->flashErrors($errors);
}
```
For me it would make more sense, since form handling does not necessarily mean that I want to save something (maybe do search, filter or even API call with form data).
It just a random idea base on this bundle https://github.com/hostnet/form-handler-bundle/wiki/Getting-Started-With-Form-Handlers what do you think?

mickaelandrieu commented:
> For me it would make more sense, since form handling does not necessarily mean that I want to save something (maybe do search, filter or even API call with form data).

Make sense, let me think a little bit about it as it may need huge work - already - to change the behavior.

Can you rebase your pull request? Sorry for the delay.

> So we could move this boilerplate code from controller to form handler?

Don't worry, in all cases, this repetitive boilerplate will be removed: I rely on interfaces to be able to improve that later without break 1.7.x branch, same apply for authorizations checks, for `legacy_controller` checks... you have some ideas here => https://github.com/PrestaShop/prestafony-project/issues/7 :)

Commit:
resolve conflict

Commit:
resolve conflict

Commit:
resolve conflict

Commit:
resolve conflict

Commit:
resolve conflict

Commit:
make form types public

Commit:
rebuild assets

sarjon commented:
> Before I didn't have the opportunity to enter negative numbers for the following fields:
> - Products per page
> - Number of days for which the product is considered 'new'
> - Display remaining quantities when the quantity is lower than

> And according to our UI Kit (https://cdn.rawgit.com/PrestaShop/prestashop-ui-kit/master/index.html#forms), can you use the custom-select class ?


@marionf fixed. :)

Commit:
use switch types instead of choice

Commit:
update stock option handler js

Commit:
rebuild assets

Commit:
use template annotation to render page

mickaelandrieu commented:
Thanks @sarjon and everyone for your kind reviews :+1:
